File: 2021/☃️November/1119/1647.cpp
//  template.cpp
//  Copyright © 2021 도찡. All rights reserved.

#ifndef _CRT_SECURE_NO_WARNINGS
#define _CRT_SECURE_NO_WARNINGS
#endif
#define _USE_MATH_DEFINES

#include <algorithm>
#include <cstring>
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;
using pii = pair<int, int>;
using lint = long long;

/* [🦔] [💜] [✨] */

int N, M;
vector<pair<int, pii>> road;

int root[100002];

void initRoot() {
    for(int i=1; i<=N; i++) root[i] = i;
}

int find(int n) {
    if(root[n] == n) return n;
    return root[n] = find(root[n]);
}

bool unionCity(int a, int b) {
    a = find(a);
    b = find(b);
    if(a == b) return false;

    root[b] = a;
    return true;
}

int main() {
#ifndef ONLINE_JUDGE
#endif

    ios_base::sync_with_stdio(false); cin.tie(0); cout.tie(0);

    cin >> N >> M;

    int a, b, c;
    for(int i=0; i<M; i++) {
        cin >> a >> b >> c;
        road.push_back({c, {a, b}});
    }

    initRoot();
    sort(road.begin(), road.end());
    int selectCount = 0;
    int ans = 0;

    for(auto &current: road) {
        int cost = current.first;
        pii cities = current.second;

        if(!unionCity(cities.first, cities.second)) continue;
        ans += cost;
        selectCount++;

        if(selectCount >= N-2) break;
    }

    cout << ans << '\n';

    return 0;
}

File: 2021/☃️November/1119/1647.swift
//  template.swift
//  Copyright © 2021 도찡. All rights reserved.

import Foundation

func readInt() -> [Int] {readLine()!.split(separator: " ").map{Int(String($0))!}}
func readStr() -> [String] {readLine()!.split(separator: " ").map{String($0)}}

typealias pii = (Int, Int)
let INF = INT32_MAX
let MOD = 1e9 + 7

/*[😈][🔮][🦔]*/

typealias Road = (first: Int, second: Int, cost: Int)

var N: Int = 0
var M: Int = 0
var road = [Road]()
var root = [Int]()

func initRoot() {
    for index in 0...N { root.append(index) }
}

func find(_ n: Int) -> Int {
    if root[n] == n { return n }
    root[n] = find(root[n])
    return root[n]
}

func union(_ a: Int, _ b: Int) -> Bool {
    let rootA = find(a);
    let rootB = find(b);
    if rootA == rootB { return false }

    root[rootB] = rootA
    return true
}

func main() {
    var input = readInt()

    N = input[0]
    M = input[1]

    for _ in 0..<M {
        input = readInt()
        road.append(Road(first: input[0], second: input[1], cost: input[2]))
    }

    initRoot()
    road.sort(by: { $0.cost < $1.cost })
    var selectCount = 0
    var ans = 0

    for current in road {
        if selectCount >= N-2 { break }
        if !union(current.first, current.second) { continue }

        ans += current.cost
        selectCount += 1
    }

    print(ans)
}
